### ISSUE TITLE
Error when importing hypergan in Python 3.6: "SyntaxError: name 'iteration' is used prior to global declaration"

### ISSUE THREAD

**Issue by subzerofun** [255BITS/HyperGAN#55]
Hey thanks for the extensive Readme and the whole project! Really looks amazing what you can do with it!

Hope you can help me out here. I've installed hypergan via 
`pip3 install hypergan --upgrade`

on OSX 10.12, python 3.6 with Tensorflow 1.0.0

My Tensorflow install is working, checked everything and even reinstalled it before upgrading hypergan – but i still get this error when trying to import it:

```
>>> import hypergan
I tensorflow/stream_executor/dso_loader.cc:135] successfully opened CUDA library libcublas.8.0.dylib locally
I tensorflow/stream_executor/dso_loader.cc:135] successfully opened CUDA library libcudnn.5.dylib locally
I tensorflow/stream_executor/dso_loader.cc:135] successfully opened CUDA library libcufft.8.0.dylib locally
I tensorflow/stream_executor/dso_loader.cc:135] successfully opened CUDA library libcuda.1.dylib locally
I tensorflow/stream_executor/dso_loader.cc:135] successfully opened CUDA library libcurand.8.0.dylib locally
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.6/site-packages/hypergan/__init__.py", line 2, in <module>
    from .gan import GAN
  File "/usr/local/lib/python3.6/site-packages/hypergan/gan.py", line 17, in <module>
    from hypergan.trainers import *
  File "/usr/local/lib/python3.6/site-packages/hypergan/trainers/adam_trainer.py", line 90
    global iteration
SyntaxError: name 'iteration' is used prior to global declaration
```

Do you know what causes this error and how i could fix it?

**Comment by martyn** [comment]
Thanks for the nice feedback!

I just released a fix for this issue.  If you upgrade with `pip3 install hypergan --upgrade` it should get further.  I haven't tried running on python 3.6 yet.  Let me know if it breaks.

**Comment by subzerofun** [comment]
Thanks for the quick answer and fix!

Looks like it works now:
```
pip3 install hypergan --upgrade
Collecting hypergan
  Downloading hypergan-0.8.5.tar.gz
Building wheels for collected packages: hypergan
  Running setup.py bdist_wheel for hypergan ... done
  Stored in directory: lalalala/pip/wheels/f6/bf/c0/c5e08657760bd2d3d70d5dff1fe15481ffbc0353d5bb74b753
Successfully built hypergan
Installing collected packages: hypergan
  Found existing installation: hypergan 0.8.4
    Uninstalling hypergan-0.8.4:
      Successfully uninstalled hypergan-0.8.4
Successfully installed hypergan-0.8.5
```

```
python3
Python 3.6.0 (default, Dec 24 2016, 08:01:42)
[GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.42.1)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import hypergan
I tensorflow/stream_executor/dso_loader.cc:135] successfully opened CUDA library libcublas.8.0.dylib locally
I tensorflow/stream_executor/dso_loader.cc:135] successfully opened CUDA library libcudnn.5.dylib locally
I tensorflow/stream_executor/dso_loader.cc:135] successfully opened CUDA library libcufft.8.0.dylib locally
I tensorflow/stream_executor/dso_loader.cc:135] successfully opened CUDA library libcuda.1.dylib locally
I tensorflow/stream_executor/dso_loader.cc:135] successfully opened CUDA library libcurand.8.0.dylib locally
>>>
```

Will try the test examples now. Can i ask something about the super-resolution.py (or training data in general)? How many pictures do i appr. need for training? Should they all be resized to 64x64px? How much can i crank up the resolution with a 6B VRAM card?

I plan to use these datasets for training: http://msvocds.blob.core.windows.net/coco2014/train2014.ziphttp://msvocds.blob.core.windows.net/coco2014/train2014.zip
and the CelebA aligned images.

When the training is finished, how can i test the model, how do i add an input image? Couldn't find a flag for that in the super-resolution.py file. Or do i load the model with the hypergan command and an input argument?

Sorry for all the questions :-/

**Comment by martyn** [comment]
No worries.  Super resolution is very beta right now.  You can train with the super-resolution.py example script.  If you want to sample or serve from it you'll need to create a server(check out 'colorizer-server').

For 6gb you should probably stay at 64x64 or 32x32.  YMMV.  I test on an 8gb card so it should be enough space.

The hypergan command doesn't load a super-resolution model right now.  I'll file a bug, since it's true for all the examples.  You can see `colorizer-server` for an example on hacking around this in 0.8.

**Comment by subzerofun** [comment]
OK, i've downloaded the mscoco and celeba dataset – do i need to prepare the files somehow? Or is it OK that they have different proportions & resolutions (between 300 to 640px width/height, mscoco images) ?

Should i use all 82.783 files for training? 😅

So when the training is finished i'll look into colorizer-server. What's the difference between all the colorizer examples?
- colorizer-builder.py
- colorizer-sampler.py
- colorizer-server.py
- colorizer.py

I'm just a little bit confused because the description for all of them is "_Train a colorizer!_".

Compared to other GAN projects you're code is very short and compact. I was trying to understand the GAN implementation (DCGAN) from this example: https://github.com/david-gpu/srez
Or this one: https://github.com/david-gpu/deep-makeover

But it's just to hard to follow... Not to say that the code is not written well – i'm just too inexperienced to understand everything.

I know this is off-topic, but the second project generates images from the celeba-dataset and lets you change the gender of a person's face. Works very well! The only problem is that the output resolution is quite low (80 x 100px).

Would it be hard to rewrite the deep-makeover code for/with hypergan? I would really love to try it out and change a few parameters, like input (training) and output resolution, etc. – but the input arguments for the generator are unfortunately very limited and the GAN model is hardcoded for just one output size. I can't figure out how i would need to modify the model to accept larger input/create larger output images. I've tried reaching out to the author, but he doesn't have the intention to add something to the code – it is just meant as an experiment and the code is finished.

Hope i don't ask for too much (sorry in advance) – but could you write a few lines, a rough sketch of how the deep-makeover project would look like with hypergan? Just as a starting point to work with.

I wouldn't mind if you're interested in adding an example for gender conversion 😎.

---

Thanks for filing the bug – i'm excited for the 0.8 release!

There is a lot of development in the GAN field – i see a new method (or improvements) every other week. So do you plan to include all the different types of GANs that will follow? That would be great – one framework to try it out all the different versions, without getting lost comparing the code from different authors.

**Comment by subzerofun** [comment]
OK i've downloaded the mscoco dataset and i already had the celeba images. Do i need to prepare the mscoco images somehow? They all have different proportions – from 300 to 640px width/height.

So i will try to run the training for super-resolution and when i have a finished model i'll look ino colorizer-server.py. Should i use all 80.000+ images for training? 😅

Could you please explain the differences of all the colorizer files?
- colorizer-builder.py
- colorizer-sampler.py
- colorizer-server.py
- colorizer.py

I'm just asking because the description is the same for all of them: _"Train a super resolution enhancer!"_

I really like that your code is clean, short and organised! I've tried to understand the DCGAN implementation of these two projects: 
https://github.com/david-gpu/srez
https://github.com/david-gpu/deep-makeover
but unfortunately i'm too inexperienced to comprehend every line of the code...

The deep-makeover project looks really interesting: It trains on the celebA dataset and can convert the gender of a persons face 🙃. Male to female and vice versa, even male to male (making people prettier) and female to female. But this repo is unfortunately only an expirement and not intended to be used as a tool (or "finished software"). It has a lot of limitations, since options like input image size (training) and output resolution are all hardcoded – the generated images are only 80 x 100px. I've tried reaching out to the author how to change the model/training files and output generator but he answered that he has no intention of extending the code. That i would have to change the model and generator myself, but i couldn't figure out how to do that...

Do you think it would be hard to rewrite the gender-conversion code for hypergan? Could you draw a rough sketch how you would structure the code – just that i get an idea on how to possibly write it myself? I don't think i'm good enough, but it would be a good training lesson.

I wouldn't mind if you're interested in adding the gender-conversion to your examples 😎 (one can always try 😄).

---

And thanks for the filing the bug. Looking forward to the 0.8 release!

Do you plan to include new GAN methods in your project? Every other day i read through new papers on [arxiv-sanity.com](http://www.arxiv-sanity.com/) or [gitxiv.com](www.gitxiv.com) and see that there is a lot of development on GANs. It would be great if your hyperGAN project becomes one framework where you can test all kinds of GANs without having to read through the code of different authors.

---

Thanks again for the quick reply!

**Comment by subzerofun** [comment]
And already the next problem :-/ – i'm such a beginner...

When i try to run
`python3 super-resolution.py /Volumes/AppData/ML-models/mscoco/*.jpg`

I get the error:
```
-bash: /usr/local/bin/python3: Argument list too long
```

Info: The github project is on my main drive (~/github/ and all my datasets are on another ssd. 
The mscoco images are in `/Volumes/AppData/ML-models/mscoco/dataset/` – 82.783 images

### **Obvious: Too many images...**
So i created a second folder with 2.000 files and tried the executing it with the following commands:
1)
`python3 super-resolution.py directory /Volumes/AppData/ML-models/mscoco-hypergan/*.jpg
`
2)
`python3 super-resolution.py --save_every 10000 --sample_every 50 -s 64x64x3 -b 32 --crop true -f jpg -d /gpu:0 /Volumes/AppData/ML-models/mscoco-hypergan/*.jpg`

But both times i've got the following error:
```
usage: super-resolution.py [-h] [--save_every SAVE_EVERY]
                           [--sample_every SAMPLE_EVERY] [--size SIZE]
                           [--batch_size BATCH_SIZE] [--crop CROP]
                           [--format FORMAT] [--device DEVICE]
                           [--use_hc_io USE_HC_IO]
                           directory
super-resolution.py: error: unrecognized arguments: /Volumes/AppData/ML-models/mscoco-hypergan/COCO_train2014_000000000025.jpg /Volumes/AppData/ML-models/mscoco-hypergan/COCO_train2014_000000000030.jpg /Volumes/AppData/ML-models/mscoco-hypergan/COCO_train2014_000000000034.jpg (... the list goes on for all 2.000 images)
```

I just tried adding all input args per hand because i thought it couldn't read the defaults.


### Copy images to the examples folder

So the next thing i tried was apparently more successfull. I copied the files over in the project folder, to `/examples/images`
and then executed
`python3 super-resolution.py -f jpg ../examples`

The network was initializing, my cards fans turned up but then... i got this error:
```bash
I tensorflow/stream_executor/dso_loader.cc:135] successfully opened CUDA library libcublas.8.0.dylib locally
I tensorflow/stream_executor/dso_loader.cc:135] successfully opened CUDA library libcudnn.5.dylib locally
I tensorflow/stream_executor/dso_loader.cc:135] successfully opened CUDA library libcufft.8.0.dylib locally
I tensorflow/stream_executor/dso_loader.cc:135] successfully opened CUDA library libcuda.1.dylib locally
I tensorflow/stream_executor/dso_loader.cc:135] successfully opened CUDA library libcurand.8.0.dylib locally
[loader] ImageLoader found 2000 images with 15 different class labels
W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE4.1 instructions, but these are available on your machine and could speed up CPU computations.
W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE4.2 instructions, but these are available on your machine and could speed up CPU computations.
W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use AVX instructions, but these are available on your machine and could speed up CPU computations.
I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:874] OS X does not support NUMA - returning NUMA node zero
I tensorflow/core/common_runtime/gpu/gpu_device.cc:885] Found device 0 with properties:
name: GeForce GTX 780
major: 3 minor: 5 memoryClockRate (GHz) 0.9015
pciBusID 0000:01:00.0
Total memory: 6.00GiB
Free memory: 1.34GiB
I tensorflow/core/common_runtime/gpu/gpu_device.cc:906] DMA: 0
I tensorflow/core/common_runtime/gpu/gpu_device.cc:916] 0:   Y
I tensorflow/core/common_runtime/gpu/gpu_device.cc:975] Creating TensorFlow device (/gpu:0) -> (device: 0, name: GeForce GTX 780, pci bus id: 0000:01:00.0)
add_lowres [4, 4]
set xfiltered Tensor("generator/ResizeNearestNeighbor:0", shape=(32, 4, 4, 3), dtype=float32, device=/device:GPU:0) [4, 4]
add_lowres [8, 8]
set xfiltered Tensor("generator/ResizeNearestNeighbor_2:0", shape=(32, 8, 8, 3), dtype=float32, device=/device:GPU:0) [8, 8]
[generator] layer Tensor("generator/g_layers_0/BiasAdd:0", shape=(32, 8, 8, 257), dtype=float32, device=/device:GPU:0) 16448
add_lowres [16, 16]
set xfiltered Tensor("generator/ResizeNearestNeighbor_4:0", shape=(32, 16, 16, 3), dtype=float32, device=/device:GPU:0) [16, 16]
[generator] layer Tensor("generator/g_layers_1/BiasAdd:0", shape=(32, 16, 16, 128), dtype=float32, device=/device:GPU:0) 32768
add_lowres [32, 32]
[generator] layer Tensor("generator/g_layers_2/BiasAdd:0", shape=(32, 32, 32, 64), dtype=float32, device=/device:GPU:0) 65536
add_lowres [64, 64]
[generator] layer Tensor("generator/g_layers_3/BiasAdd:0", shape=(32, 64, 64, 3), dtype=float32, device=/device:GPU:0) 12288
[discriminator] layer Tensor("discriminator/AvgPool:0", shape=(64, 32, 32, 32), dtype=float32, device=/device:GPU:0)
[discriminator] layer Tensor("discriminator/AvgPool_1:0", shape=(64, 16, 16, 70), dtype=float32, device=/device:GPU:0)
[discriminator] layer Tensor("discriminator/AvgPool_2:0", shape=(64, 8, 8, 146), dtype=float32, device=/device:GPU:0)
[discriminator] layer Tensor("discriminator/AvgPool_3:0", shape=(64, 4, 4, 292), dtype=float32, device=/device:GPU:0)
[discriminator] layer Tensor("discriminator/AvgPool_4:0", shape=(64, 2, 2, 584), dtype=float32, device=/device:GPU:0)
 |= Initializing new network
W tensorflow/core/common_runtime/bfc_allocator.cc:217] Ran out of memory trying to allocate 1.20GiB. The caller indicates that this is not a failure, but may mean that there could be performance gains if more memory is available.
W tensorflow/core/common_runtime/bfc_allocator.cc:217] Ran out of memory trying to allocate 946.32MiB. The caller indicates that this is not a failure, but may mean that there could be performance gains if more memory is available.
W tensorflow/core/common_runtime/bfc_allocator.cc:217] Ran out of memory trying to allocate 946.32MiB. The caller indicates that this is not a failure, but may mean that there could be performance gains if more memory is available.
W tensorflow/core/common_runtime/bfc_allocator.cc:217] Ran out of memory trying to allocate 1.20GiB. The caller indicates that this is not a failure, but may mean that there could be performance gains if more memory is available.
 0: g cost 0.16 d_loss 1.22 d_real -1.20 d_log -0.20
Sampling 50
-- (32, 64, 64, 3)
bwxshape (32, 64, 64, 3) (64, 64, 3)
Traceback (most recent call last):
  File "super-resolution.py", line 119, in <module>
    gan.sample_to_file(sample_file, sampler=sampler)
  File "/usr/local/lib/python3.6/site-packages/hypergan/gan.py", line 42, in sample_to_file
    return sampler(self, name)
  File "super-resolution.py", line 44, in sampler
    plot(config, images, name)
  File "/usr/local/lib/python3.6/site-packages/hypergan/samplers/common.py", line 7, in plot
    imsave(file, image)
  File "/usr/local/lib/python3.6/site-packages/scipy/misc/pilutil.py", line 197, in imsave
    im.save(name)
  File "/usr/local/lib/python3.6/site-packages/PIL/Image.py", line 1725, in save
    fp = builtins.open(filename, "w+b")
FileNotFoundError: [Errno 2] No such file or directory: 'samples/50.png'
```

### What does this error mean?

```
FileNotFoundError: [Errno 2] No such file or directory: 'samples/50.png'
```

What did i do wrong this time 😅 ?

**Comment by martyn** [comment]
Appreciate the writeup :)  Try running again after `mkdir samples`.

**Comment by martyn** [comment]
I might miss some of your questions but I'll try to answer:

In regards to plans for HyperGAN - we're going to support major developments quickly.

In regards to the deep-makeover - it should be possible to implement with HG.  I haven't read enough of that code to understand how it's implemented tho.  The easiest way - use class loss.  Place your 'male' and 'female' photos in different directories, then run hypergan.  Switch the Y(class) during sampling.  I'm not sure the manifolds would align, but they might.

As for image resizing - by default they are scaled without preserving scale.  You can switch to cropping instead by the `--crop` option.

Gotta run, I think I missed a bunch of questions tho.

**Comment by subzerofun** [comment]
OK good to know – reading through the code and i'm slowly getting the hang of it.

Thanks for the info about how to structure the gender conversion code – will definitely try it out!

I finally got the super resolution code running, my GPU is now training with 2.000 images (hope that's enough). Have you tested out what's a good amount of images – for super res?

Unfortunately the `--crop` option didn't work... I think i ~might~ create a new issue and post the error code there. So that other users can benefit from finding the solution. I think this thread is already long enough 😄.

But thanks again for all the quick responses!

Uhm the part you missed was this:

---
**Could you please explain the differences of all the colorizer files?**

- colorizer-builder.py
- colorizer-sampler.py
- colorizer-server.py
- colorizer.py

I'm just asking because the description is the same for all of them: "_Train a super resolution enhancer!_"

---

**Comment by martyn** [comment]
Thanks, I didn't realize `--crop` was broken.  It looks like a TF 1.0 issue.  I can take a look.

In regards to the different colorizer files - they are hacks.  :)   It's basically because the cli doesn't work with those api examples yet.  You `colorizer-server` can be used instead of `hypergan serve`, etc.

As for the number of images - it depends.  The more the better.  My test training set has 200k faces.

**Comment by subzerofun** [comment]
Thanks for the info, just a quick result of my first run of super-resolution.py:

It generated 199 pngs (last sample nr. was 4975) and then crashed (yippie!) because of a missing folder – it tried to save the checkpoint file to a non-existing folder:
`~/.hypergan/**saves**`

Error code:
```
"Parent directory of {} doesn't exist, can't save.".format(save_path))
ValueError: Parent directory of /Users/[username]/.hypergan/saves/super-resolution.ckpt doesn't exist, can't save.
```

So i just had to create the `saves` folder and now checkpoint saving is working!

You just mentioned that your training set has 200k faces – did you mean the celebA dataset (202.600 images) ?

But how did you manage to load that many files – didn't you get the same error as me with the mscoco dataset?
`-bash: /usr/local/bin/python3: Argument list too long`

What do i have too look out for to avoid that error? I'll try running super res again, with the full mscoco dataset (82.000 files). But it will take a while to copy the images to the hypergan project folder on my main SSD...

What would be the correct command for running the super res training when i want to use files from a different volume?
Most of my datasets are on `/Volumes/AppData/ML-models/`

So when i want to load the mscoco files, which are saved here: `/Volumes/Appdata/ML-models/mscoco-hypergan` what would be the correct argument to include the folder?
`python3 super-resolution.py -f jpg [pleasefillthisspace]`

I've tried
`python3 super-resolution.py -f jpg /Volumes/AppData/ML-models/mscoco-hypergan/*.jpg`
before, but it didn't work (got the `Argument list too long` error...).

**Comment by martyn** [comment]
The cli argument should point to folder instead.  So 

```bash
  python3 super-resolution.py -f jpg /Volumes/Appdata/ML-models/mscoco-hypergan
```

You may need to place all of your jpg files into a subfolder(for example /Volumes/Appdata/ML-models/all/)
